#17 Documentation

msimet opened the pull request:
Here's a draft of the Stile documentation!  This isn't intended to be a finished pull request for you all to just sign off on--please edit anything you don't think is clear.

This PR required changing the style of a lot of docstrings, so the diff is huge.  The things that really changed are:
- Any file in the `doc/` directory, all of which are new;
- A bit of the `HistogramSysTest`, which added a `field=` kwarg to be in line with the way we do the `StatSysTest`;
- A new checklist for PRs in `devel/checklist.txt`;
- A couple of small fixes in `stile/__init__.py` and `tests/test_binning.py`.

To build the documentation, make sure you have Sphinx installed and that Stile and all its dependencies are in your `$PYTHONPATH`, then type:

```
make html
```

to generate HTML docs, or

```
make latexpdf
```

to make a PDF.  (I think the HTML version looks better.)  `make clean html` will rebuild all the files if you want that; otherwise it will only regenerate the documentation for any files that changed.

There are some intro files (one each for systematics tests, data structure, and binning), then a bunch of `.rst` files read by Sphinx that contain automatically-generated documentation (those are all the "automodule" commands.)  I'm happy to answer questions here about Sphinx formatting things if you want to change something but don't know how.

rmandelb commented:
I installed sphinx and was able to make the html dox, but when I tried the PDF, I got
```
! LaTeX Error: File `newfloat.sty' not found.
```

Are you doing this on coma?  Is there some place with a latex distribution that has more stuff than the default?  (I made the PDF successfully on my Mac, but I don't have the lsst stack installed there so I get errors when doing the html build.)

On a side note - I notice the README still says
```
Stile is a pure python package, no compilation needed. We do not currently have a utility that will install it in your system path for you, but as long as this directory is in your $PYTHONPATH, you should be able to import and run.
```
but we do have setup.py, so I think this can be updated...

msimet commented:
Ah--I hadn't run it on coma, actually.  There's no newfloat.sty on there as far as I can tell.  

There's actually an issue about this on the Sphinx repo now--some Linux distributions don't include this file, so they'll  be packaging it with future Sphinx versions just in case.  The solution for now is to download newfloat.sty (I found a copy at http://math.sut.ac.th/lab/software/texlive/texmf-dist/tex/latex/caption/newfloat.sty ) and put it in a directory ~/texmf/tex/latex/.  Then try running

    kpsewhich newfloat.sty

and if that shows the file, latex compilation will work.  How annoying!

I've updated the README on this branch, thanks.

rmandelb commented:
Yep, that works.  Thanks.

rmandelb commented:
Some basic, silly comments on the PDF:  The TOC page says CONTENTS at the upper right, and has a page number (i).  Then I see a completely blank page without header/footer, then a page numbered 1 that says "Content:" on it (only that, plus header/footer), then a blank page 2 that does have a header+footer.   So I think something weird is happening to the introductory material.

rmandelb commented:
I know the README says this stuff, but I think the intro should say stuff about dependencies and installation, or if you don't want to worry about keeping the files in sync, you could say "see the README in the repo for information about dependencies and installation."

rmandelb commented on doc/systest_basics.rst:
```diff
@@ -0,0 +1,288 @@
+catalogs are given, they are used as random lenses, and
+:math:`\langle \gamma_t \rangle-\langle \gamma_{t, {\rm rand}} \rangle` is returned instead.
+
+Shear-shear correlation functions are :math:`\xi_+` and :math:`\xi_-`.  :math:`\xi_+` is,
```
Honestly, I understood this because I understood these definitions already, but I'm not sure a newbie would get it.  Perhaps just say "the definitions of xi+ and xi- follow the formalism of BLAH" (where BLAH is one of the many papers that defines this xi+ / xi- formalism) and not try to do this in words?

rmandelb commented on doc/systest_basics.rst:
```diff
@@ -0,0 +1,288 @@
+from the two catalogs, not the two components of a single shear in sky coordinates or chip frame.)
+
+Point-scalar (point-kappa) estimates are equivalent to <scalar>; scalar-shear estimates are
+equivalent to <scalar*Re(shear)> with a corresponding imaginary case; scalar-scalar estimates
```
So perhaps say explicitly that in this case, the real (imaginary) case is the tangential (45 degree) shear of the 2nd sample about the positions defined by the 1st sample?

rmandelb commented on doc/systest_basics.rst:
```diff
@@ -0,0 +1,288 @@
+also an optional comparison trendline at :math:`y=0` or :math:`x=y` (or a user-defined function).
+The predefined types are:
+
+- **StarVsPSFG1**: star vs PSF g1
```
Really g1 / g2?  Or t and x, as in the text a bit higher up?

msimet replied:
Yeah, g1/g2--this is just comparing estimates of the same objects so there's no obvious direction to choose as t and x, unlike for the spatial correlations.  Should we highlight this?

rmandelb replied:
No, it's OK.

rmandelb commented on doc/systest_basics.rst:
```diff
@@ -0,0 +1,288 @@
+based on two different algorithms:
+
+>>> scott_rule_hist = stile.HistogramSysTest(binning_style='scott')  # Scott's rule
+>>> freedman_rule_hist = stile.HistogramSysTest(binning_style='freedman')  # Freedman-Diaconis rule
```
If it's not too annoying, can you say briefly what these rules are?  I'm not familiar with them, and I think if we can write a single sentence that prevents people running off to google, that would be helpful.

rmandelb commented on doc/systest_basics.rst:
```diff
@@ -0,0 +1,288 @@
+Currently, a :class:`stile.StatSysTest <stile.sys_tests.StatSysTest>` will compute min, max,
+median, median absolute deviation (MAD), mean, standard deviation, variance, and number of
+objects, and will additionally compute the skew and kurtosis if ``scipy.stats`` can be imported.
+It will also compute percentiles in the data: by default, 1, 2, and 3 sigma values around the
```
Perhaps clarify that this is based on the confidence intervals for a Gaussian - i.e., it won't compute sigma and then use that to tell you the 1/2/3-sigma values, but rather uses the numpy percentiles routine to get the values of the data at the percentiles corresponding to 1, 2, and 3 sigma about the median.

rmandelb commented on doc/data.rst:
```diff
@@ -0,0 +1,62 @@
+- **ra**, the RA of the object
+- **x**, the x coordinate of the object
+- **y**, the y coordinate of the object
+- **g1**, a shear component in the ra or x direction
```
Perhaps a few words about what this "or" means.

rmandelb commented on doc/binning_basics.rst:
```diff
@@ -0,0 +1,102 @@
+---------------
+
+To start binning your data, you create a :class:`Bin*` object that will contain a binning
+definitions.  For instance, if you wanted to bin the ``ra`` column in 10 bins:
```
"a binning definition" or "binning definitions"

rmandelb commented on doc/binning_basics.rst:
```diff
@@ -0,0 +1,102 @@
+========================
+
+Stile contains some simple functions to bin your data.  Two of them--:class:`stile.BinStep
+<stile.binning.BinStep>` and :class:`stile.BinList <stile.binning.BinList>`--have simple,
```
Could use some context for how/what this is used for -- i.e., which systematics tests might require binning.

rmandelb commented on doc/binning_basics.rst:
```diff
@@ -0,0 +1,102 @@
+produce properly binned subsets.
+
+>>> for single_bin in bin_object():
+>>>     binned_data = single_bin(data)
```
Perhaps clarify what ```binned_data``` is; is it a subset of the data in the same format as ```data``` itself, does the format change at all, ...?

rmandelb commented on doc/binning_basics.rst:
```diff
@@ -0,0 +1,102 @@
+knows its boundaries and it also contains a string you can use in program outputs.
+
+>>> single_bin = bin_object()[0]
+>>> type(single_bin)
```
I think there's a line missing after this one where it outputs the type?   But you already said what it was above, so maybe this line is just not needed here at all?

rmandelb commented on doc/binning_basics.rst:
```diff
@@ -0,0 +1,102 @@
+with a list of bin edges and a field name (see the :doc:`data` documentation for more information
+on field names).
+
+>>> bin_list_object = stile.BinList(field='g1', bin_list=[-10, -1, -0.5, -0.3, -0.1, -0.05, 0, 0.05, 0.1, 0.3, 0.5, 1, 10])
```
Is there a way to get a line-break so it doesn't go off the side of the document in the PDF?

rmandelb commented on doc/binning_basics.rst:
```diff
@@ -0,0 +1,102 @@
+if it returns Boolean masks.  This object can be called like any other :class:`Bin*` object to
+create a list of callable objects, and it will work with :func:`stile.ExpandBinList
+<stile.binning.ExpandBinList>` as well.  However, the child objects it creates when you call it
+don't have ``.low`` or ``.high`` attributes!
```
Say what are the practical implications of this.

rmandelb commented on doc/conf.py:
```diff
@@ -0,0 +1,287 @@
+#  author, documentclass [howto, manual, or own class]).
+latex_documents = [
+  (master_doc, 'Stile.tex', u'Stile Documentation',
+   u'Melanie Simet, Hironao Miyatake, and the Stile development team', 'manual'),
```
Author list runs off side of page in PDF.  Put line break in it?

rmandelb commented on doc/codedoc.rst:
```diff
@@ -0,0 +1,13 @@
+==================
+Code documentation
```
Would it make sense to have a few words here like "here are the docstrings for the main classes that users will see..." before jumping straight into the docstrings?

msimet commented:
Hi folks, I think I have successfully addressed all the comments from...um...a year ago.  If anyone has time to look over the changes and see if they make sense, that would be great!

rmandelb commented on doc/systest_basics.rst:
```diff
@@ -0,0 +1,297 @@
+- **StarVsPSFSigma**: star vs PSF sigma
+- **ResidualVsPSFG1**: (star - PSF) g1 vs PSF g1
+- **ResidualVsPSFG2**: (star - PSF) g1 vs PSF g2
+- **ResidualVsPSFSigma**: (star - PSF) g1 vs PSF sigma
```
I'm a little confused why these are all g1 residual vs. (other stuff) but e.g. we don't have g2 residual vs. g2 or sigma residual vs. sigma.  Is that a typo?

msimet replied:
Yep, it's a typo.  Fixed.

rmandelb commented:
Looks good to me.  I say go ahead and merge.
